### DISCUSSION TITLE
[@types/react] typing of `useRef(null)` allows for runtime errors

### DISCUSSION THREAD

**Discussion by atomanyih** [DefinitelyTyped/DefinitelyTyped#55783] [Issues with a @types package]
tl;dr: `useRef(null)` creates a ref that can be passed to any function that expects a `RefObject`

## Problem

Came across this the other day when someone forgot to pass a type parameter to a standard element ref.

```tsx
import React, { useRef, RefObject } from 'react';

const useSomething = (ref: RefObject<{doSomething(): void}>) => {
  useEffect(() => {
    if(ref.current) {
      ref.current.doSomething()
    }
  }, []};
};

const What = () => {
  const ref = useRef(null);
  useSomething(ref);

  return <div ref={ref}/>;
};
```
[typescript playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDecCuAzksgGbZkDyARgFZIZwC+cpUEIcA5FKhlwG4AUELQQAdoXhEkAZQ5IYAC2DiA5nAC8cABS9SALkRJSNehgA8OACYR5IRSvU6AlEYBuEYNaYA+F1q+uEJwcMCkeiYAdGj4ULziMAE4IaFw+jFxCTBRtvaOqmquqUxCpaISUnAA6koo8NqugcGhYpLw+loExGQ64vgANgMuwqEy+cqFkaQjIqG8MHHicBbWwO7pJpo4+kwA9L7C5Xt73SQmfYPDcADuwMoQ+PAocDAAnmBIcCji4hAw9WAEjChFeHyQ1m+oIAsk8UNQBudTHQGDALP0hr4hCc4NQntg0D84FJgENcV8wChCMRITAIN9lkhEQ5EiIcT83pN1K86vAAQBrJCgqgoywAFSCdzJ6DQSDA8FhAIRSLMqPRV18QA)

The above code passes typecheck, but at runtime the ref will be populated with a `HTMLDivElement` and the custom hook will try to call a method that does not exist and throw an error.

However, typescript is totally okay with this. `useRef(null)` returns a `MutableRefObject<null>`, which can be passed to any function that expects a `RefObject`. (`RefObject<T>` is `{current: T | null}`, which is satisfied by `{current: null}` 
 (`MutableRefObject<null>`) (_or at least this is what I think is happening_)

I'm not sure if this is intended, but for us it is a bit of a foot gun, as a dev can easily create a ref that bypasses some amount of type-checking.

## Solution?

In our codebase we amend the React type definitions with this:
```ts
declare namespace React {
  function useRef(initialValue: null): RefObject<null>;
}
```
So the above code will desirably fail with this error:
```
TS2345: Argument of type 'RefObject<null>' is not assignable to parameter of type 'RefObject<{doSomething(): void}>'.
  Type 'null' is not assignable to type '{doSomething(): void}'.
```

Is this kosher?

[a comment by an automated account is not shown]

**Comment by atomanyih** [comment]
This hasn't received any answer in ~3 months. Is github discussions still the right place for package-specific issues?
